README.md
# ng-vega [![NPM version][npm-image]][npm-url]

Angular directive for rendering [Vega](http://vega.github.io/) specs.
This project was forked and modified from [angular-vega](https://github.com/eptify/angular-vega) which was written for Vega 1 and became inactive.
The current version of ng-vega supports Vega 2.

### Demo

- [Simple demo](http://kristw.github.io/ng-vega) -- Select dataset/renderer to see the chart changes and see the [code](https://github.com/kristw/ng-vega/blob/master/examples/index.html) to see how it was implemented.
- [Vega editor demo](http://kristw.github.io/ng-vega/editor.html) -- Implement Vega editor using ng-vega.

For more information about Vega, please refer to [official documentation](http://trifacta.github.io/vega/).

### Usage

```javascript
angular.module('exampleApp', ['ngVega'])
```

```html
<div vega spec="spec" vega-data="testData" vega-renderer="'svg'" vega-on-parse="myCallback(view)"></div>
```

- `spec` is `$scope.spec` in your controller.

- `vega-data` (optional) can be used to pass dynamic data. In the example above, it is bound to `$scope.testData`. Data can be a function to modify the values (Vega 2 syntax) or raw values (and ng-vega will convert it to a function to make it work for you).

```javascript
$scope.testData = {
  // function to modify dataset name "table"
  table: function(data){
    data.remove(function(d){return true;})
      insert([{a: 3}, {a: 4}])
  }
}

$scope.testData = {
  // raw values for dataset name "table"
  table: [{a: 1},{a: 2}]
}
```

- `vega-renderer` (optional) can be used to set renderer (`'canvas'` or `'svg'`). Don't forget the quote.

- `vega-on-parse` (optional) can be used to get notified when `spec` is parsed and to receive a handle to the [View Component](https://github.com/vega/vega/wiki/Runtime#view-component-api) that Vega has built for the corresponding chart. In the example above, this parameter is bound to `$scope.myCallback`.

```javascript
$scope.myCallback = function(view) {
  // perform any work you need to do once the spec is parsed
  console.log('Vega spec has been parsed.');
  // make use of the View Component API as you wish
  view.on('click', function(event, item) {
    console.log('clicked on ' + JSON.stringify(item));
  });
};
```

### Installation

```
bower install ng-vega --save
```

or

```
npm install ng-vega --save
```

### Import into your project

Angular module `ngVega` will be available once you do one of the following:

##### Choice 1. Global

Adding this library via ```<script>``` tag is the simplest way.

```html
<script src="path/to/angular.js"></script>
<script src="path/to/vega.js"></script>
<script src="path/to/ng-vega.min.js"></script>
```

##### Choice 2: AMD

If you use requirejs, this library support AMD out of the box.

```javascript
require.config({
  paths: {
    angular:   'path/to/angular',
    vega:      'path/to/vega',
    'ng-vega': 'path/to/ng-vega'
  }
});
require(['ng-vega'], function() {
  // do something
});
```

##### Choice 3: node.js / browserify

```javascript
require('ng-vega');
```

### Author

Krist Wongsuphasawat / [@kristw](https://twitter.com/kristw)


Copyright (c) 2016 Krist Wongsuphasawat. MIT License

[npm-image]: https://badge.fury.io/js/ng-vega.svg
[npm-url]: https://npmjs.org/package/ng-vega


CHANGELOG.md
# ng-vega

## v1.4.0
- Add `vega-on-parse` attribute to the `vega` directive. See more information from pull request [#3](https://github.com/kristw/ng-vega/pull/3). Thanks @[mhuyck](https://github.com/mhuyck)

## v1.3.1-1.3.2
- Add unnecessary files to ignore list for bower and npm.

## v1.3.0
- Add Vega 2 as dependency
- Support new syntax for setting data via the [runtime view API](https://github.com/vega/vega/wiki/Upgrading-to-2.0) with backward compatibility for setting raw values.

## v1.2.0
- Remove vega from dependency


editor.js
angular.module('exampleEditor', ['ngVega', 'ui.ace'])
  .controller('EditorController', function($scope) {
    $scope.renderer = 'canvas';

    $scope.specString = JSON.stringify({
      "width": 300,
      "height": 200,
      "padding": {"top": 10, "left": 30, "bottom": 30, "right": 10},
      "data": [
        // only describe a name for dataset here and load data dynamically
        {
          "name": "table",
          "values": [
            {"x": 1,  "y": 280}, {"x": 2,  "y": 55},
            {"x": 3,  "y": 43}, {"x": 4,  "y": 91},
            {"x": 5,  "y": 81}, {"x": 6,  "y": 53},
            {"x": 7,  "y": 19}, {"x": 8,  "y": 87},
            {"x": 9,  "y": 52}, {"x": 10, "y": 48},
            {"x": 11, "y": 24}, {"x": 12, "y": 49},
            {"x": 13, "y": 87}, {"x": 14, "y": 66},
            {"x": 15, "y": 17}, {"x": 16, "y": 27},
            {"x": 17, "y": 68}, {"x": 18, "y": 16},
            {"x": 19, "y": 49}, {"x": 20, "y": 15}
          ]
        }
      ],
      "scales": [
        {
          "name": "x",
          "type": "ordinal",
          "range": "width",
          "domain": {"data": "table", "field": "x"}
        },
        {
          "name": "y",
          "range": "height",
          "nice": true,
          "domain": {"data": "table", "field": "y"}
        }
      ],
      "axes": [
        {"type": "x", "scale": "x"},
        {"type": "y", "scale": "y"}
      ],
      "marks": [
        {
          "type": "rect",
          "from": {"data": "table"},
          "key": "x",
          "properties": {
            "enter": {
              "x": {"scale": "x", "field": "x"},
              "width": {"scale": "x", "band": true, "offset": -1},
              "y": {"scale": "y", "field": "y"},
              "y2": {"scale": "y", "value": 0}
            },
            "update": {
              "x": {"scale": "x", "field": "x"},
              "width": {"scale": "x", "band": true, "offset": -1},
              "y": {"scale": "y", "field": "y"},
              "y2": {"scale": "y", "value": 0},
              "fill": {"value": "steelblue"}
            },
            "hover": {
              "fill": {"value": "red"}
            }
          }
        }
      ]
    }, null, 2);

    $scope.$watch('specString', function(value){
      try{
        $scope.spec = JSON.parse(value);
        $scope.errorMessage = false;
      }
      catch(ex){
        $scope.errorMessage = ex.message;
      }
    });

  });


src/ng-vega.js
// Define module using Universal Module Definition pattern
// https://github.com/umdjs/umd/blob/master/returnExports.js

(function (root, factory) {
  if (typeof define === 'function' && define.amd) {
    // Support AMD. Register as an anonymous module.
    define(['angular', 'vega'], factory);
  }
  else if (typeof exports === 'object') {
    // Node/CommonJS
    module.exports = factory(require('angular'), require('vega'));
  }
  else {
    // Just define it in angular and done
    factory(root.angular, root.vg);
  }
}(this, function (angular, vg) {
  //---------------------------------------------------
  // BEGIN code for this module
  //---------------------------------------------------

  function debounce(func, wait) {
    var timeout;

    return function() {
      var context = this, args = arguments;
      var later = function() {
        timeout = null;
        func.apply(context, args);
      };
      clearTimeout(timeout);
      timeout = setTimeout(later, wait);
      // return caller for chaining
      return context;
    };
  }

  return angular.module('ngVega', [])
    .directive('vega', function() {
      return {
        restrict: 'AE',
        scope: {
          spec: '=',
          data: '=vegaData',
          renderer: '=vegaRenderer',
          onParse: '&vegaOnParse'
        },
        link: function(scope, elements, attrs) {
          var element = elements[0];
          var view;
          var processedData;

          function parse(){
            vg.parse.spec(scope.spec, function(chart) {
              view = chart({
                el: element,
                data: processedData,
                renderer: scope.renderer || 'svg'
              }).update();
              scope.onParse({view: view});
            });
          }

          var debouncedParse = debounce(parse, 50);

          scope.$watch('spec', debouncedParse, true);

          scope.$watch('data', function(data){
            processedData = {};

            if(angular.isDefined(data)){
              Object.keys(data).forEach(function(key){
                var value = data[key];
                processedData[key] = angular.isFunction(value) ? value : function(dat){
                  dat.remove(function(d){return true;})
                    .insert(value);
                };
              });
            }

            if(view){
              view.data(processedData).update();
            }
            else{
              debouncedParse();
            }
          }, true);

          scope.$watch('renderer', function(renderer){
            if(view){
              view.renderer(renderer).update();
            }
            else{
              debouncedParse();
            }
          });
        }
      };
    });

  //---------------------------------------------------
  // END code for this module
  //---------------------------------------------------
}));

app.js
var example = angular.module('exampleApp', ['ngVega']);

example.controller('BarChartController', function($scope) {
  $scope.renderer = 'canvas';
  $scope.spec = {
    "width": 300,
    "height": 200,
    "padding": {"top": 10, "left": 30, "bottom": 30, "right": 10},
    "data": [
      // only describe a name for dataset here and load data dynamically
      {"name": "table"}
    ],
    "scales": [
      {
        "name": "x",
        "type": "ordinal",
        "range": "width",
        "domain": {"data": "table", "field": "x"}
      },
      {
        "name": "y",
        "range": "height",
        "nice": true,
        "domain": {"data": "table", "field": "y"}
      }
    ],
    "axes": [
      {"type": "x", "scale": "x"},
      {"type": "y", "scale": "y"}
    ],
    "marks": [
      {
        "type": "rect",
        "from": {"data": "table"},
        "key": "x",
        "properties": {
          "enter": {
            "x": {"scale": "x", "field": "x"},
            "width": {"scale": "x", "band": true, "offset": -1},
            "y": {"scale": "y", "field": "y"},
            "y2": {"scale": "y", "value": 0}
          },
          "update": {
            "x": {"scale": "x", "field": "x"},
            "width": {"scale": "x", "band": true, "offset": -1},
            "y": {"scale": "y", "field": "y"},
            "y2": {"scale": "y", "value": 0},
            "fill": {"value": "steelblue"}
          },
          "hover": {
            "fill": {"value": "red"}
          }
        }
      }
    ]
  };

  var data1 = {
    table: [
      {"x": 1,  "y": 280}, {"x": 2,  "y": 55},
      {"x": 3,  "y": 43}, {"x": 4,  "y": 91},
      {"x": 5,  "y": 81}, {"x": 6,  "y": 53},
      {"x": 7,  "y": 19}, {"x": 8,  "y": 87},
      {"x": 9,  "y": 52}, {"x": 10, "y": 48},
      {"x": 11, "y": 24}, {"x": 12, "y": 49},
      {"x": 13, "y": 87}, {"x": 14, "y": 66},
      {"x": 15, "y": 17}, {"x": 16, "y": 27},
      {"x": 17, "y": 68}, {"x": 18, "y": 16},
      {"x": 19, "y": 49}, {"x": 20, "y": 15}
    ]
  };
  var data2 = {
    table: [
      {"x": 1,  "y": 2}, {"x": 2,  "y": 55},
      {"x": 3,  "y": 1}, {"x": 4,  "y": 91},
      {"x": 5,  "y": 81}, {"x": 6,  "y": 53},
      {"x": 7,  "y": 19}, {"x": 8,  "y": 87},
      {"x": 9,  "y": 52}, {"x": 10, "y": 48},
      {"x": 11, "y": 24}, {"x": 12, "y": 49},
      {"x": 13, "y": 87}, {"x": 14, "y": 66},
      {"x": 15, "y": 17}, {"x": 16, "y": 27}
    ]
  };

  $scope.testData = data1;
  $scope.data1 = data1;
  $scope.data2 = data2;
});
